thecourseforum/theCourseForum2
# Fix grades

# Status
Done

# What I did
- No big changes in the code except discarding middle names, but creating a PR for the SQL scripts to be merged ASAP.
- Will continue to improve `load_grades.py` in the coming weeks

## Issues addressed
- This is a quick and dirty fix to #293 but not a permanent one. However, running `db/link_courseinstructorgrade_to_instructor.sql` will allow us to restore around 80% of the missing `CourseInstructor` instances by giving it a proper `Instructor` foreign key.

## Screenshots
![image](https://user-images.githubusercontent.com/7676354/101262125-94b20100-377f-11eb-9c8c-c1dbd9d38886.png)
4034 -> 834 is roughly an 80% decrease in the number of cases where `CourseInstructor` was not linked to `Instructor`.

## How to Test 
- Start by running `docker exec -ti tcf_db psql tcf_db -U tcf_django -c "SELECT first_name, last_name, course_id, COUNT(*) AS num FROM tcf_website_courseinstructorgrade WHERE course_id IS NOT NULL GROUP BY first_name, last_name, course_id ORDER BY num DESC;"` (and then exit by entering `\q` + return). You should see a few rows whose `num` columns are not 1 on the top.
- Run `docker-compose down && docker-compose up`, wait until the Django server is up and running, and then run the following commands in a separate terminal, one at a time in the following order:
```
docker exec -i tcf_db psql tcf_db -U tcf_django -c "DROP SCHEMA public CASCADE; CREATE SCHEMA public;"
docker exec -i tcf_db psql tcf_db -U tcf_django < launch.sql
// In the production environment, we will start from here, creating a backup beforehand to be safe.
// Of course, we will only run the part after `docker exec -i {containerName}` since we don't use Docker on GAE or DO.
// The SQL scripts are not in the DO, so you can copy/paste the content instead of using the `-f` flag.
docker exec -i tcf_django python3 manage.py load_grades ALL_DANGEROUS 
docker exec -i tcf_django python3 manage.py fix_multiple_objects_returned
docker exec -i tcf_db psql tcf_db -U tcf_django -f app/link_courseinstructorgrade_to_instructor.sql
docker exec -i tcf_db psql tcf_db -U tcf_django -f app/add_missing_instructor_emails.sql
```
- Now everything _should_ be fixed. Again, run `docker exec -ti tcf_db psql tcf_db -U tcf_django -c "SELECT course_id, instructor_id, COUNT(*) AS num FROM tcf_website_courseinstructorgrade cig WHERE course_id IS NOT NULL AND instructor_id IS NOT NULL GROUP BY course_id, instructor_id ORDER BY num DESC;"` (and then entering `q` + return). You should only see 1 in the `num` column now.
